keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1234 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding [field selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/)
+support to custom resources.
+
+## Motivation
+
+All Kubernetes APIs support [field
+selection](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/)
+for metadata.name, metadata.namespace.
+
+Additionally, built-in APIs support field selection on specific fields such as
+status.phase for Pods, but CustomResourceDefinitions lack equivalent
+functionality.
+
+Without this enhancement, an extension author that wants to be be able to filter
+resources might choose to:
+
+  - Organize data into a label even though it should be a spec or status field.
+  - Double writing data both to a label and a spec or status field.
+
+Both of these should be avoided. We should enable extension authors structure
+their APIs according to best practices, not based on data access patterns.
+
+### Goals
+
+- Add an allow list of selectable fields to CustomResourceDefinitions.
+- Performance and resource consumption characteristics are roughtly equivalent
+  to label selectors.
+
+### Non-Goals
+
+- Arbitrary field selection.
+   - This proposal is to support field selection for a small allow list of
+     fields. Expanding support to include all fields is complicated by:
+     - Fields contained in, or nested under, lists or maps.
+     - The costs to apiserver of evaluating filters on objects or maintaining
+       indices.
+     - Lack of direct support for filtering in etcd.
+- Support for conversions, casts or other transforms of field values beyond
+  the automatic cast to string required for field selection.
+- More sophisticated field selection logic. Possible approaches include: CEL
+  based field selection, enhancements to field selection such as set based
+  operators like provided for labels ("in", "not in"), OR clauses, pattern
+  matching... See the "CEL as an accelerated predicate language" section "Future
+  Work" for more details on what makes this challenging.
+
+
+## Proposal
+
+Add `selectableFields` to the versions of CustomResourceDefinition:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: selector.stable.example.com
+spec:
+  group: stable.example.com
+  versions:
+    - name: v1
+      selectableFields:
+      - jsonPath: .spec.color 
+      - jsonPath: .spec.size 
+      additionalPrinterColumns:
+      - jsonPath: .spec.color
+        name: Color
+        type: string
+      - jsonPath: .spec.size
+        name: Size
+        type: string
+...
+```
+
+(Alternatively, `selectableFields` could contain `fieldPath: spec.color` to
```

Comment:
question: Is there a reason selectableFields contains jsonPath while the OpenAPI output uses fieldPath?